src/momento/aio/_add_header_client_interceptor.py

```diff
@@ -32,13 +32,51 @@ async def intercept_unary_unary(
         client_call_details: grpc.aio._interceptor.ClientCallDetails,
         request: grpc.aio._typing.RequestType,
     ) -> Union[grpc.aio._call.UnaryUnaryCall, grpc.aio._typing.ResponseType]:
+
+        # FIXME Hack to account for ddtrace lib injecting grpc metadata as list instead of proper grpc.aio.Metadata()
+        # Will open issue on data dog repo to discuss this more.
+        # Since once client_call_details.metadata is set as list we cant overwrite with `Metadata()` class we build
+        # up a new ClientCallDetails() in each request now in this interceptor and make sure Metadata() is set properly
+        new_client_call_details = None
+
+        # If no metadata set on passed in client call details then we are first to set, so we should just initialize
         if client_call_details.metadata is None:
-            client_call_details.metadata = Metadata()
+            new_client_call_details = ClientCallDetails(
+                method=client_call_details.method,
+                timeout=client_call_details.timeout,
+                metadata=Metadata(),
+                credentials=client_call_details.credentials,
+                wait_for_ready=client_call_details.wait_for_ready,
+            )
+
+        # This is block hit when ddtrace interceptor runs first and sets metadata as a list
+        elif isinstance(client_call_details.metadata, list):
+            existing_headers = client_call_details.metadata
+            metadata = Metadata()
+            # re-add all existing values to new metadata
+            for md_key, md_value in existing_headers:
+                metadata.add(md_key, md_value)
+            new_client_call_details = ClientCallDetails(
+                method=client_call_details.method,
+                timeout=client_call_details.timeout,
+                metadata=metadata,
+                credentials=client_call_details.credentials,
+                wait_for_ready=client_call_details.wait_for_ready,
+            )
+        else:
+            # Default assuming previous interceptors have passed proper `Metadata()` object for now and just copy over
+            new_client_call_details = ClientCallDetails(
+                method=client_call_details.method,
+                timeout=client_call_details.timeout,
+                metadata=client_call_details.metadata,
+                credentials=client_call_details.credentials,
+                wait_for_ready=client_call_details.wait_for_ready,
```

Comment:
This logic is hairy enough that I think it should be extracted into a helper function or helper class so that unit tests can be written about it, which will make assertions about how the final `ClientCallDetails` look given various flavors of `Metadata`-ish objects that are being passed into it.

If you really need to unblock the customer and you are 100% sure this works, I am okay with you merging and then addressing in a follow-up commit, but the code that handles the datadog path looks suspect to me at first glance, and I think this has to be a like "today or tomorrow" follow-up PR, not like a "put it in the backlog" type of follow-up PR.  let me know if you disagree with that.

Reply:
ack let me throw in function and write some tests for it quickly I really want to just come back in and delete interceptor completely but agree I think should do this since could linger here for longer then I hope.

Reply:
pulled out and added unit tests